Convert GenerateServerKeys to GRPC

zmb3 opened the pull request:
Updates #6394

zmb3 committed:
Convert GenerateServerKeys to GRPC

Updates #6394

zmb3 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1567,6 +1573,7 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 		return nil, trace.Wrap(err)
 	}
 
+	// TODO(zmb3): add CA certs?
```
Reviewers - what do you think? Any harm in including the CA certs when we generate user certs just so it behaves similarly to when we generate host certs?

codingllama replied:
Sounds alright.

zmb3 commented on lib/auth/auth.go:
```diff
@@ -1554,67 +1554,24 @@ func HostFQDN(hostUUID, clusterName string) string {
 	}
 
-	if err := a.limiter.AcquireConnection(req.Roles.String()); err != nil {
+	reqRoles, err := types.NewTeleportRoles(req.Roles)
```
Note: now that the request is `[]string` instead of `types.SystemRoles` we do a little extra validation here.

zmb3 commented on lib/auth/auth.go:
```diff
@@ -1889,21 +1846,27 @@ func (a *Server) RegisterUsingToken(req RegisterUsingTokenRequest) (*PackedKeys,
 	}
 	log.Infof("Node %q [%v] has joined the cluster.", req.NodeName, req.HostID)
-	return keys, nil
+	return &PackedKeys{
```
The only difference at this point between `auth.PackedKeys` and `proto.Certs` is that `PackedKeys` has a field for the private key. We don't populate that field here since the auth server doesn't have the private key, so I'm looking for thoughts on whether we should just return `proto.Certs` here as well.

codingllama replied:
If we never return the private keys, then returning proto.Certs seems to be more appropriate.

zmb3 commented on lib/auth/helpers.go:
```diff
@@ -878,18 +879,25 @@ func NewServerIdentity(clt *Server, hostID string, role types.SystemRole) (*Iden
-	keys.Key = priv
-	return ReadIdentityFromKeyPair(keys)
+
+	return ReadIdentityFromKeyPair(&PackedKeys{
```
If we change the signature to `ReadIdentityFromKeyPair(privateKey []byte, certs *proto.Certs)` we could remove `auth.PackedKeys` entirely. Thoughts?

codingllama replied:
I'm not sure I've looked at this enough to have a strong opinion. Do you mean we could remove the PackedKeys type entirely? What are we trying to achieve by doing that?

zmb3 replied:
Yes, we could remove the type entirely. Prior to this change, we had:

- GenerateUserCerts, which is in the GRPC server and returns `proto.Certs` (an SSH and TLS certificate)
- GenerateServerKeys, which did not have a GRPC endpoint but returned `PackedKeys` (an SSH and TLS certificate, the corresponding private key, as well as the CA certificates that signed them)

With this change, we update `proto.Certs` to include the CAs, making it essentially the same type as `PackedKeys`.
This is the only remaining use of `PackedKeys`, so we're converting `proto.Certs` to `PackedKeys` just so that we can call `ReadIdentityFromKeyPair`.

By removing`PackedKeys` we could reduce some potential confusion for having two similar types, and also avoid the situation we have today where `PackedKeys` usually has a nil private key field but sometimes doesn't.

codingllama replied:
+1 for removing PackedKeys. Thanks for the explanation.

stevenGravy committed:
Expanded AWS Console examples (#8127)

* Expanded AWS Console docs with more examples

* Corrected file names for cert file

stevenGravy committed:
Expand error message on tctl enterprise usage (#8093)

tcsc committed:
Port backend tests to testify / fix racy tests (#8170)

In order to get better visibility into the backend database tests using
the standard Go tooling, this changeset ports the backend tests away
from `Check`, and into subtests & `testify` for assertions.

This change means that individual sub-tests
 1. can be more easily identified in the json test logs, and
 2. can be more easily run individually from the command line

During this port I also discovered that some tests are using the fake
clocks incorrectly, which may be a cause of some of our flaky etcd
tests.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -150,6 +150,40 @@ message Certs {
+        [ (gogoproto.jsontag) = "additional_principals,omitempty" ];
+    // DNSNames is a list of DNS names to include in x509 certificates.
+    repeated string DNSNames = 5 [ (gogoproto.jsontag) = "dns_names,omitempty" ];
+    // PublicTLSKey is a PEM encoded public key used for TLS setup,
```
```suggestion
    // PublicTLSKey is a PEM encoded public key used for TLS setup.
```

codingllama commented on api/client/proto/types.go:
```diff
@@ -0,0 +1,14 @@
+	if req.HostID == "" {
+		return trace.BadParameter("missing parameter HostID")
+	}
+	if len(req.Roles) != 1 {
```
Should this be `<1`?  If not then why do we take a list?

zmb3 replied:
For host certs we always have a single role (node, auth, proxy, etc.)

I'm not sure why the request takes a slice, it seems to have always been that way. This change is not new code, it's just moving `CheckAndSetDefaults` from the old `auth.GenerateServerKeysRequest` to the new `protocol.GenerateServerKeysRequest`.

codingllama replied:
Right. Maybe leave a TODO to clean this later? Seems strange.

zmb3 replied:
I will just change this to a single role and not a list. Easier to do now than later.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -150,6 +150,40 @@ message Certs {
+        [ (gogoproto.jsontag) = "additional_principals,omitempty" ];
+    // DNSNames is a list of DNS names to include in x509 certificates.
+    repeated string DNSNames = 5 [ (gogoproto.jsontag) = "dns_names,omitempty" ];
+    // PublicTLSKey is a PEM encoded public key used for TLS setup,
+    bytes PublicTLSKey = 6 [ (gogoproto.jsontag) = "public_tls_key" ];
+    // PublicSSHKey is a SSH encoded public key.
+    bytes PublicSSHKey = 7 [ (gogoproto.jsontag) = "public_ssh_key" ];
```
Are the Public TLS / Public SSH the keys signed? Are they optional? Do they have to match? If they have to match, can't we just generate the SSH key from the TLS one?

Please document the answers.

zmb3 replied:
They are not yet signed. This message is essentially a CSR - you send the auth server some public keys and it sends back signed certificates.

They are both required. We used to support omitting them (both) and allowing the auth server to generate the private key and corresponding public keys, but that behavior never actually worked outside of test code due to a bug and was recently removed in #8135. 

In all cases I'm aware of, clients generate these from the same private key, but the auth server doesn't know or cares whether or not that is true  - it's just going to return signed certificates to the client.

I will take a stab at adding this detail to the proto message.

codingllama replied:
Right, thanks.

codingllama commented on lib/auth/auth.go:
```diff
@@ -1685,7 +1642,7 @@ func (a *Server) GenerateServerKeys(req GenerateServerKeysRequest) (*PackedKeys,
 	}
 
-	isAdminRole := req.Roles.Equals(types.SystemRoles{types.RoleAdmin})
+	isAdminRole := reqRoles.Equals(types.SystemRoles{types.RoleAdmin})
```
Do a contains check instead of an equals?

zmb3 replied:
When it comes to deciding whether or not to grant admin privileges I'm inclined to not change the existing behavior in any way without strong reason to.

codingllama replied:
Fair enough.

zmb3 committed:
Remove auth.PackedKeys in favor of proto.Certs

zmb3 committed:
Merge branch 'master' into zmb3/convert-generate-server-keys-grpc

zmb3 committed:
Add missing license header

zmb3 committed:
Include the CAs when generating user certs as well

zmb3 commented on lib/auth/clt.go:
```diff
@@ -2009,7 +1991,7 @@ type ClientI interface {
 	// GenerateServerKeys generates new host private keys and certificates (signed
 	// by the host certificate authority) for a node
-	GenerateServerKeys(GenerateServerKeysRequest) (*PackedKeys, error)
+	GenerateServerKeys(context.Context, *proto.GenerateServerKeysRequest) (*proto.Certs, error)
```
I’m in favor. Never liked the name as it’s not generating keys at all, it’s generating certs.

zmb3 commented on lib/auth/httpfallback.go:
```diff
@@ -798,3 +798,30 @@ func (c *Client) ChangeUserAuthentication(ctx context.Context, req *proto.Change
+
+// GenearteServerKeys generates new host certificates (signed by the host CA).
+func (c *Client) GenerateServerKeys(ctx context.Context, req *proto.GenerateServerKeysRequest) (*proto.Certs, error) {
+	if err := req.CheckAndSetDefaults(); err != nil {
```
Wouldn’t that prevent it from running the validation in the not implemented case where it falls back to the web API?

Joerger replied:
No, because the validation logic would still run on the client side before getting an unimplemented error from the server

zmb3 committed:
Combine some small API client types into a single file

zmb3 committed:
Address review feedback

This commit contains 2 changes:

1. Rename GenerateServerKeys to GenerateHostCerts.
   This is a more accurate name and consistent with the existing
   GenerateUserCerts endpoint.
2. Change the request type to include a single role, rather than a
   list of roles. We only ever allowed a single role in the list
   anyway, so this change will prevent future mis-use of the API.

Note: a side effect of this change is we now have two similar endpoints:
- GenerateHostCert: old API that generates SSH cert only
- GenerateHostCerts: a newer API that generates SSH and TLS certs

To avoid making this change too big, we'll aim to deprecate
GenerateHostCert in the future.

zmb3 commented on lib/auth/clt.go:
```diff
@@ -577,24 +577,6 @@ func (c *Client) RegisterUsingToken(req RegisterUsingTokenRequest) (*PackedKeys,
 
-// RenewCredentials returns a new set of credentials associated
-// with the server with the same privileges
-func (c *Client) GenerateServerKeys(req GenerateServerKeysRequest) (*PackedKeys, error) {
```
I see that, and with the rename you suggested there's definitely remove for improvement here. I agree we should deprecate that old endpoint in a future PR. I

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -150,6 +150,42 @@ message Certs {
+    // NodeName is a user-friendly host name.
+    string NodeName = 2 [ (gogoproto.jsontag) = "node_name" ];
+    // Roles is a list of roles assigned to the host.
+    repeated string Roles = 3 [ (gogoproto.jsontag) = "roles" ];
```
Thanks! `(gogoproto.casttype) = "github.com/gravitational/teleport.api/types.SystemRole"` did the trick

zmb3 committed:
Merge branch 'master' into zmb3/convert-generate-server-keys-grpc

zmb3 commented:
@Joerger @codingllama I believe I've addressed all your feedback, PTAL.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -156,15 +156,18 @@ message Certs {
+// HostCertsRequest specifies certificate-generation parameters
 // for a server.
-message GenerateServerKeysRequest {
+message HostCertsRequest {
```
nit: GenerateHostCertsRequest, so it matches the RPC name?

```suggestion
message GenerateHostCertsRequest {
```

zmb3 replied:
We already have `rpc GenerateUserCerts` that takes a `UserCertsRequest` so I was trying to be consistent with existing RPCs.

codingllama replied:
Consistency is not our strong suite (other parts of the gRPC API go for RPC Foo / FooRequest / FooResponse) - I stand by my suggestion, as I think it's the better practice, but up to you.

Joerger replied:
I agree with this change.

codingllama reviewed (approved):
LGTM.

Joerger commented on api/client/proto/types.go:
```diff
@@ -14,13 +14,28 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package proto provides protobuf api specification for Teleport Auth servers and clients.
```
Add the godoc package description to types.go.

zmb3 committed:
Merge branch 'master' into zmb3/convert-generate-server-keys-grpc

zmb3 committed:
Add godoc for package proto

zmb3 committed:
Merge branch 'master' into zmb3/convert-generate-server-keys-grpc

russjones reviewed (approved):
Bot.
